dev/RadioButtons/RadioButtons.cpp

```diff
@@ -7,136 +7,634 @@
+                                    {
+                                        auto const newFocusedRepeaterIndex = repeater.GetElementIndex(newFocusedElementAsUIElement);
+                                        if ((oldFocusedRepeaterIndex == 0 && newFocusedRepeaterIndex == repeaterItemCount - 1) ||
+                                            (oldFocusedRepeaterIndex == repeaterItemCount - 1 && newFocusedRepeaterIndex == 0))
+                                        {
+                                            args.TrySetNewFocusedElement(oldFocusedElement);
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
     }
+}
 
-    UpdateItemsSource();
+void RadioButtons::OnRepeaterLoaded(const winrt::IInspectable&, const winrt::RoutedEventArgs&)
+{
+    if (auto const repeater = m_repeater.get())
+    {
+        if (m_testHooksEnabled)
+        {
+            AttachToLayoutChanged();
+        }
+
+        UpdateSelectedIndex();
+        UpdateSelectedItem();
+        UpdateMaximumColumns();
+        OnRepeaterCollectionChanged(nullptr, nullptr);
+    }
+}
+
+void RadioButtons::KeyDownHandler(const winrt::IInspectable&, const winrt::KeyRoutedEventArgs& args)
+{
+    if (args.Key() == winrt::VirtualKey::Control)
+    {
+        m_isControlDown = true;
+    }
+}
+
+void RadioButtons::KeyUpHandler(const winrt::IInspectable&, const winrt::KeyRoutedEventArgs& args)
+{
+    if (args.Key() == winrt::VirtualKey::Control)
+    {
+        m_isControlDown = false;
+    }
 }
 
-void RadioButtons::OnListViewLoaded(const winrt::IInspectable& /*sender*/, const winrt::RoutedEventArgs& args)
+void RadioButtons::OnChildKeyDown(const winrt::IInspectable&, const winrt::KeyRoutedEventArgs& args)
 {
-    if (auto listView = m_listView.get())
+    if (args.OriginalKey() == winrt::VirtualKey::Down)
     {
-        m_itemsWrapGrid.set(SharedHelpers::FindInVisualTreeByType<winrt::ItemsWrapGrid>(listView));
-        if (auto itemsWrapGrid = m_itemsWrapGrid.get())
+        if (MoveFocusNext())
         {
-            // Override normal up/down behavior -- down should always go to the next item and up to the previous.
-            m_listViewKeyDownRevoker = itemsWrapGrid.KeyDown(winrt::auto_revoke, { this, &RadioButtons::OnListViewKeyDown });
-            m_listViewKeyUpRevoker = itemsWrapGrid.KeyUp(winrt::auto_revoke, { this, &RadioButtons::OnListViewKeyUp });
+            args.Handled(true);
+        }
+    }
+    else if (args.OriginalKey() == winrt::VirtualKey::Up)
+    {
+        if (MoveFocusPrevious())
+        {
+            args.Handled(true);
+        }
+    }
+    else if (args.OriginalKey() == winrt::VirtualKey::Right)
+    {
+        if (MoveFocusRight(args.OriginalSource().try_as<winrt::UIElement>()))
```

Comment:
For Right and Left, you could possibly avoid a lot of your custom logic by using XYFocusNavigation instead.
https://docs.microsoft.com/en-us/uwp/api/Windows.UI.Xaml.Input.FocusManager.FindNextFocusableElement

Reply:
yay !

---
In reply to: [347572820](https://github.com/microsoft/microsoft-ui-xaml/pull/1623#discussion_r347572820) [](ancestors = 347572820)